rfcs/react-components/styles-handbook.md

```diff
@@ -0,0 +1,771 @@
+
+function App(props) {
+  const classes = useClasses();
+  const className = mergeClasses(classes.base, props.primary && classes.primary);
+
+  /* --- */
+}
+```
+
+### Avoid `!important`
+
+Our styles are written in way to allow predictable and simple style overrides, `!important` should not be used in styles:
+
+```js
+import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
+
+const useClasses = makeStyles({
+  base: {
+    // ❌ Don't do
+    display: 'flex !important',
+  },
+});
+```
+
+## Performance
+
+### Avoid nested `mergeClasses` calls
+
+`mergeClasses` is a performant function, however it's not expected that it will be called multiple times for the same element.
+
+```js
+// ❌ Don't do
+function Component(props) {
+  const classes = useClasses();
+
+  const classesForFoo = mergeClasses(/* ---- */);
+  const className = mergeClasses(classes.root, classesForFoo, mergeClasses(/* ---- */), mergeClasses(/* ---- */));
+
+  /* --- */
+}
+```
+
+Conditions to apply styles might be complex, in this case consider to extract them to separate variables:
+
+```js
+// ✅ Do
+function Component(props) {
+  const classes = useClasses();
+
+  const conditionForFoo = /* ---- */ true;
+  const className = mergeClasses(classes.root, conditionForFoo && classes.foo /* other condition */);
+
+  /* --- */
+}
+```
+
+## Nested selectors
+
+Nested selectors in styles should be used responsibly as they can cause "CSS rules explosion" problem.
+
+### Usage with pseudo classes
+
+Use nested selectors when they are combined with pseudo classes:
+
+```js
+import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
+import { tokens } from '@fluentui/react-theme';
+
+const useClasses = makeStyles({
+  base: {
+    // ✅ Do
+    // Shows filled icon on hover
+    ':hover': {
+      [`& .${iconFilledClassName}`]: {
+        display: 'inline',
+      },
+      [`& .${iconRegularClassName}`]: {
+        display: 'none',
+      },
+    },
+  },
+});
+```
+
+### Do not use selectors to target elements
+
+Do not use nested selectors to target slots or nested elements that are accessible directly:
+
+```jsx
+import { makeStyles, mergeClasses, shorthands } from '@griffel/react';
+import { tokens } from '@fluentui/react-theme';
+
+const useClasses = makeStyles({
+  root: {
+    backgroundColor: tokens.colorNeutralBackground1,
+    // ❌ Don't do
+    // You can apply classes directly to that "div"
+    '> div': {
+      color: tokens.colorNeutralForeground1,
+    },
+  },
+
+  // ✅ Do
+  root: {
+    backgroundColor: tokens.colorNeutralBackground1,
+  },
+  slot: {
+    color: tokens.colorNeutralForeground1,
+  },
+});
+
+function App(props) {
+  const classes = useClasses();
+
+  return (
+    <div className={classes.root}>
+      {/* 💡 Classes can be passed directly to this element */}
+      <div className={classes.slot} />
+    </div>
+  );
+}
+```
+
+### Do not introduce complicated selectors
+
+Keep selectors simple to produce reusable CSS rules:
```

Comment:
let's lso add that complicated selectors are hard to override

Reply:
Added in 18de44f.